File: kt/src/main/kotlin/de/arcan/arclib/crypto/Hkdf.kt
package de.arcan.arclib.crypto

object Hkdf {
    private const val HASH_LEN = 32

    fun deriveKey(ikm: ByteArray, salt: ByteArray?, info: ByteArray?, length: Int): ByteArray {
        val prk = extract(salt ?: ByteArray(HASH_LEN), ikm)
        return expand(prk, info ?: ByteArray(0), length)
    }

    fun extract(salt: ByteArray, ikm: ByteArray): ByteArray {
        return Hmac.sha256(salt, ikm)
    }

    fun expand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        val n = (length + HASH_LEN - 1) / HASH_LEN
        require(n <= 255) { "HKDF expand: output too large" }

        val output = ByteArray(length)
        var prev = ByteArray(0)
        var offset = 0

        for (i in 1..n) {
            val input = prev + info + byteArrayOf(i.toByte())
            prev = Hmac.sha256(prk, input)
            val take = minOf(HASH_LEN, length - offset)
            prev.copyInto(output, offset, 0, take)
            offset += take
        }
        return output
    }
}

File: js/src/hkdf.js
// RFC 5869 HKDF-SHA256.
//
// Two-step construction: Extract pseudorandom key from input keying
// material + salt, then Expand into the requested output length using
// counter-mode HMAC.

import { hmacSha256 } from './hmac.js';
import { concat } from './utils.js';

const HASH_LEN = 32;

// HKDF-Extract(salt, IKM) -> PRK
export function hkdfExtract(salt, ikm) {
  if (!salt || salt.length === 0) salt = new Uint8Array(HASH_LEN);
  return hmacSha256(salt, ikm);
}

// HKDF-Expand(PRK, info, L) -> OKM
export function hkdfExpand(prk, info, length) {
  const N = Math.ceil(length / HASH_LEN);
  if (N > 255) throw new Error('hkdf: requested output too large (>255 * HashLen)');
  if (!info) info = new Uint8Array(0);
  const out = new Uint8Array(length);
  let prev = new Uint8Array(0);
  let off = 0;
  for (let i = 1; i <= N; i++) {
    const t = hmacSha256(prk, concat(prev, info, new Uint8Array([i])));
    const take = Math.min(HASH_LEN, length - off);
    out.set(t.subarray(0, take), off);
    off += take;
    prev = t;
  }
  return out;
}

// One-shot HKDF-SHA256.
export function hkdf(salt, ikm, info, length) {
  return hkdfExpand(hkdfExtract(salt, ikm), info, length);
}
